File: unity/Assets/_Project/Scripts/Core/Rules/StartingPlayerRule.cs
#nullable enable
using System.Collections.Generic;

namespace Pose.Core
{
    /// <summary>
    /// Determines who leads the opening turn and which tile they must play.
    /// The standard rule across every variant Pose supports: the player holding
    /// the highest double leads with that tile (so [6|6] beats [5|5] beats … beats
    /// [0|0]). If no player holds any double, the player holding the single tile
    /// with the highest pip total leads with that tile, ties broken by the order
    /// players appear in the players list.
    /// </summary>
    public static class StartingPlayerRule
    {
        public readonly struct Lead
        {
            public PlayerId Player { get; }
            public Tile Tile { get; }

            public Lead(PlayerId player, Tile tile)
            {
                Player = player;
                Tile = tile;
            }
        }

        /// <summary>
        /// Finds the starting player and their leading tile. <paramref name="maxPip"/>
        /// is the maximum pip value present in the variant's tile set (6 for double-six,
        /// 9 for double-nine, etc.) and bounds the double-search range.
        /// </summary>
        public static Lead FindLead(
            IReadOnlyList<PlayerId> players,
            IReadOnlyDictionary<PlayerId, Hand> hands,
            byte maxPip)
        {
            for (int d = maxPip; d >= 0; d--)
            {
                Tile target = new((byte)d, (byte)d);
                for (int i = 0; i < players.Count; i++)
                {
                    PlayerId p = players[i];
                    if (hands[p].Contains(target))
                    {
                        return new Lead(p, target);
                    }
                }
            }

            // No player holds any double; fall back to the highest single tile.
            PlayerId bestPlayer = players[0];
            Tile bestTile = default;
            int bestPips = -1;
            for (int i = 0; i < players.Count; i++)
            {
                PlayerId p = players[i];
                foreach (Tile t in hands[p])
                {
                    if (t.Pips > bestPips)
                    {
                        bestPips = t.Pips;
                        bestPlayer = p;
                        bestTile = t;
                    }
                }
            }

            return new Lead(bestPlayer, bestTile);
        }
    }
}

File: functions/src/rules/startingPlayerRule.ts
import { Hand } from './hand';
import { PlayerId } from './ids';
import { Tile } from './tile';

/** Who leads and with which tile. Port of `StartingPlayerRule.Lead`. */
export interface Lead {
  readonly player: PlayerId;
  readonly tile: Tile;
}

/**
 * Finds the opening lead. Port of `Pose.Core.StartingPlayerRule.FindLead`: the
 * holder of the highest double leads with it; if nobody holds a double, the
 * holder of the highest-pip single tile leads, ties broken by player order.
 *
 * `hands` is parallel to `players` (index i is player i's hand).
 */
export function findLead(
  players: readonly PlayerId[],
  hands: readonly Hand[],
  maxPip: number,
): Lead {
  for (let d = maxPip; d >= 0; d--) {
    const target = new Tile(d, d);
    for (let i = 0; i < players.length; i++) {
      const hand = hands[i];
      const player = players[i];
      if (hand !== undefined && player !== undefined && hand.contains(target)) {
        return { player, tile: target };
      }
    }
  }

  // No doubles anywhere — fall back to the highest single tile.
  const firstPlayer = players[0];
  if (firstPlayer === undefined) {
    throw new Error('findLead requires at least one player.');
  }
  let bestPlayer = firstPlayer;
  let bestTile = new Tile(0, 0);
  let bestPips = -1;
  for (const [i, player] of players.entries()) {
    const hand = hands[i];
    if (hand === undefined) {
      continue;
    }
    for (const t of hand.tiles) {
      if (t.pips > bestPips) {
        bestPips = t.pips;
        bestPlayer = player;
        bestTile = t;
      }
    }
  }
  return { player: bestPlayer, tile: bestTile };
}
